Auto-update the current project on buffer switch, file open etc

katspaugh opened the pull request:
This is a feature mentioned in #22.

@sabof please tell me if you prefer that instead of a separate command `project-explorer-start`, we add a custom boolean variable `pe/auto-update` and start auto-updating on `project-explorer-open`.

katspaugh commented on project-explorer.el:
```diff
@@ -1736,6 +1737,37 @@ outside of the project's root."
+  (remove-hook 'post-command-hook #'pe/post-command-update))
+
+;;;###autoload
+(cl-defun project-explorer-auto ()
```
Just `defun` might be OK?

sabof commented:
Emacs has something called `auto-revert-mode`. Might it be better suited?

katspaugh commented:
@sabof not sure if we are on the same page. What this PR does is updates the project-explorer sidebar to show the project corresponding to the opened file.

sabof commented:
Oh, I see. I'll test and review this at some point.

katspaugh commented:
Great, thanks!

By the way, I could only run the tests after changing `tree-equal` to `cl-tree-equal`.

dgutov commented:
I think you can use `window-configuration-change-hook` here. Should be a bit more reliable than `post-command-hook` and a timer.

sabof commented:
It doesn't seem to run when selecting a different window.

dgutov commented:
You're right, sorry. No better suggestion here.

sabof commented:
Here are some comments:

- It becomes impossible to close the side-bar.
- Probably new buffers shouldn't be automatically created -- accidentally starting a scan on a very large directory becomes too easy.
- Could the variable be renamed to something like `pe/follow-current`? `pe/auto-update` seems misleading.
- I'm not sure that the extra functions are needed -- I'd imagine it's an option most people would set once.
- `pe/quit` isn't the only way do delete the buffer -- perhaps you could check for the variable in the hook, and disable it if needed.

katspaugh commented:
> It becomes impossible to close the side-bar.

Fixed.

> Could the variable be renamed to something like pe/follow-current? pe/auto-update seems misleading.

Yep, renamed.

> pe/quit isn't the only way do delete the buffer -- perhaps you could check for the variable in the hook, and disable it if needed.

Done. I check in the timer callback if `(pe/get-project-explorer-window)` is opened.

> Probably new buffers shouldn't be automatically created -- accidentally starting a scan on a very large directory becomes too easy.

Agreed, I'll add the check and will also squash all commits into one. Will let you know. Thanks!

katspaugh commented:
@sabof I added a check to avoid scanning directory. It will either reuse existing buffers or load the cache if the cache file exists. Please review once more.

sabof commented:
Seems to work. 

- I think you could have re-used `project-explorer-open`, instead of creating a separate `pe/open`
- `pe/get-cache-file-name` is also very similar to `pe/cache-make-filename`
- "update" is still used in several places related to the feature.

katspaugh commented:
I've renamed the vars and edited some docs to only use words like `follow` and `switch`.
`pe/get-cache-file-name` – renamed to `pr/project-cache-filename`. Is it OK not to start a function with a verb?

> I think you could have re-used project-explorer-open, instead of creating a separate pe/open

`project-explorer-open` is opening the project explorer window *and* starts listening on the post-command hook. `pe/open` just opens the window.

sabof commented:
You could have used `pe/goto-file`.

sabof commented:
When I said `pe/get-cache-file-name` is very similar to `pe/cache-make-filename`, I meant what it does is very similar. It doesn't have a reason to exist.

katspaugh commented:
> You could have used pe/goto-file.

I tried calling `pe/goto-file` instead of `pe/open`, but it doesn't create new project buffers from cache. I also noticed `pe/show-file` called `project-explorer-open`, so I replaced it with `pe/open` as well.

> When I said pe/get-cache-file-name is very similar to pe/cache-make-filename, I meant what it does is very similar. It doesn't have a reason to exist.

`pe/project-cache-filename`'s purpose is very different. It returns the existing cache file path, and only if it exists.

sabof commented:
Another option would be to check `pe/follow-current` in the hook, and run `pe/stop-follow-current` from there. 

Valid "reasons to exist" include the following, and I don't see how `pe/project-cache-filename` qualifies:

- It wraps significant logic
- Reading the name conveys the intent more clearly
- It significantly reduces code duplication.

sabof commented:
Or even better, leave the hook as is, and call 
```
(when pe/follow-current
  (project-explorer-open))
```
instead of `pe/open`. `pe/show-file` is perfectly fine using the existing `project-explorer-open`

sabof commented:
Actually, you are already doing this, meaning that putting `project-explorer-open` there would work exactly the same.

katspaugh commented:
@sabof sorry, I don't think I understand what you mean by the last 2 comments.

The situation as I see it:

 * We need to start listening to the hook when the user calls a specific function (we can't do it just by requiring the file).
 * I propose using `project-explorer-open` as such function so as not to inflate the API.
 * Thus, inner code shouldn't anymore call `project-explorer-open`, to avoid setting multiple listeners.
 * The inner code, when it needs to (re)display the project, should call `pe/open` instead. Examples of callers of `pe/open`: `pe/show-file`, `pe/follow-current-open`.

Please correct the points you think I'm wrong about.

Speaking of `pe/project-cache-filename`, I added it to reduce duplication which you mentioned as a valid reason to create a function. This function is used twice–in `pe/follow-current-open` and in `pe/cache-load`.

sabof commented:
`add-hook` doesn't add the same listener multiple times.

**significantly**.

katspaugh commented:
OK, got it!

I've accidentally closed this PR by resetting my branch, will open a new one.
